mne/preprocessing/otp.py

```diff
@@ -0,0 +1,141 @@
+    individual sensor noise under the assumption that sampled fields
+    (e.g., MEG and EEG) are oversampled by the sensor array [1]_.
+
+    OTP can improve sensor noise levels (especially under visual
+    inspection) and repair some bad channels. This noise reduction is known
+    to interact with :func:`tSSS <mne.preprocessing.maxwell_filter>` such
+    that increasing the ``st_correlation`` value will likely be necessary.
+
+    Channels marked as bad will not be used to reconstruct good channels,
+    but good channels will be used to process the bad channels. Depending
+    on the type of noise present in the bad channels, this might make
+    them usable again.
+
+    Use of this algorithm is covered by a provisional patent.
+
+    .. versionadded:: 0.16
+
+    References
+    ----------
+    .. [1] Larson E, Taulu S (2017). Reducing Sensor Noise in MEG and EEG
+           Recordings Using Oversampled Temporal Projection.
+           IEEE Transactions on Biomedical Engineering.
+    """
+    logger.info('Processing MEG data using oversampled temporal projection')
+    if picks is None:
+        picks = _pick_data_channels(raw.info, exclude=())
+    picks_good, picks_bad = list(), list()
+    for pi in picks:
+        if raw.ch_names[pi] in raw.info['bads']:
+            picks_bad.append(pi)
+        else:
+            picks_good.append(pi)
+    del picks
+    picks_good = np.array(picks_good, int)
+    picks_bad = np.array(picks_bad, int)
+
+    n_samp = int(round(float(duration) * raw.info['sfreq']))
+    starts, stops, windows = _get_lims_COLA(
+        n_samp, len(raw.times), raw.info['sfreq'], picks_good)
+    min_samp = (stops - starts).min()
+    if min_samp < len(picks_good) - 1:
+        raise ValueError('duration (%s) yielded %s samples, which is fewer '
+                         'than the number of channels -1 (%s)'
+                         % (n_samp / raw.info['sfreq'], min_samp,
+                            len(picks_good) - 1))
+    raw_orig = raw.copy()
+    raw = raw.copy().load_data(verbose=False)
+    raw._data[picks_good] = 0.
+    raw._data[picks_bad] = 0.
+    for start, stop, window in zip(starts, stops, windows):
+        logger.info('    Denoising % 8.2f - % 8.2f sec'
+                    % tuple(raw.times[[start, stop - 1]]))
+        data_picked = raw_orig[picks_good, start:stop][0]
+        if not np.isfinite(data_picked).all():
+            raise RuntimeError('non-finite data (inf or nan) found in raw '
+                               'instance')
+        # demean our slice and our copy
+        data_picked_means = np.mean(data_picked, axis=-1, keepdims=True)
+        data_picked -= data_picked_means
+        # scale the copy that will be used to form the temporal basis vectors
+        # so that _orth_svdvals thresholding should work properly with
+        # different channel types (e.g., M-EEG)
+        norms = _normalize_vectors(data_picked)
```

Comment:
can you compute norms 3liens below as you don't use them before the for ?

Reply:
no, `_normalize_vectors` operates inplace so they are actually used (already divided out by the call itself)